src/custom_hdrs.c

```diff
@@ -0,0 +1,114 @@
+#include <re.h>
+#include <baresip.h>
+#include "core.h"
+
+
+static void hdr_destructor(void *arg)
+{
+	struct sip_hdr *hdr = arg;
+
+	/*
+	list_unlink(&hdr->le);
+	hash_unlink(&hdr->he);
+	*/
+	mem_deref((char *)hdr->name.p);
+	mem_deref((char *)hdr->val.p);
+}
+
+static void custom_hdrs_destructor(void *arg)
+{
+	struct list *hdrs = arg;
+	list_flush(hdrs);
+}
+
+int custom_hdrs_alloc(struct list **hdrs)
+{
+	struct list *h;
+	h = mem_zalloc(sizeof(*h), custom_hdrs_destructor);
+	if (!h)
+		return ENOMEM;
+	list_init(h);
+	*hdrs = h;
+	return 0;
+}
+
+int custom_hdrs_add_pl(struct list *hdrs,
+					   const struct pl *name,
+					   const struct pl *val)
+{
+	int err;
+	struct sip_hdr *hdr;
+
+	hdr = mem_zalloc(sizeof(*hdr), hdr_destructor);
+	if (!hdr)
+		return ENOMEM;
+
+	err = pl_dup(&hdr->name, name);
+	if (err)
+		goto err;
+
+	err = pl_dup(&hdr->val, val);
+	if (err)
+		goto err;
+
+	hdr->id = SIP_HDR_NONE;
+
+	list_append(hdrs, &hdr->le, hdr);
+
+	return 0;
+
+err:
+	mem_deref(hdr);
+	return err;
+}
+
+int custom_hdrs_add(struct list *hdrs,
+					const char *name,
+					const char *val)
+{
+	struct pl name_pl;
+	struct pl val_pl;
+
+	pl_set_str(&name_pl, name);
+	pl_set_str(&val_pl, val);
+
+	return custom_hdrs_add_pl(hdrs, &name_pl, &val_pl);
+}
```

Comment:
I am wondering if we should add formatted arguments here, similar to:

```
int sip_treplyf(struct sip_strans **stp, struct mbuf **mbp, struct sip *sip,
		const struct sip_msg *msg, bool rec_route, uint16_t scode,
		const char *reason, const char *fmt, ...)
{
	va_list ap;
	int err;

	va_start(ap, fmt);
	err = vreplyf(stp, mbp, true, sip, msg, rec_route, scode, reason,
		      fmt, ap);
	va_end(ap);

	return err;
}
```

in this case we only need one add function, which becomes very flexible and powerful.

instead of custom_hdrs_add_int you can then write:

`custom_hdrs_add(hdrs, "X-Header", "%d", 42);`

what do you think ?